### PR TITLE
fix #14165 #18739

### LINKED ISSUE #6269: Nested calls to same iterator fail with "var not init"
I tried to run the following code
```
import future

iterator combinations*(pool: seq[char], r: int): seq[char] =
  # combinations(@['A','B','C','D'], 2) --> AB AC AD BC BD CD
  block it:
    let n = pool.len
    if r > n:
      break it
    var indices = lc[i | (i <- 0..<r), int]
    yield lc[pool[i] | (i <- indices), char]
    while true:
      var i = r - 1
      block forElse:
        while i >= 0:
          if indices[i] != i + n - r:
            break forElse
          i -= 1
        break it
      indices[i] += 1
      for j in i+1..<r:
        indices[j] = indices[j-1] + 1
      yield lc[pool[i] | (i <- indices), char]

for s1 in combinations(@['A','B','C','D'], 2):
  echo s1
  for s2 in combinations(@['1','2','3','4'], 2):
    echo s1, " ", s2
  
```
but it fails with
```
example.nim(3, 10) Error: internal error: expr: var not init :env_110863
No stack traceback available
To create a stacktrace, rerun compilation with ./koch temp c <file>
```
However, only one loop runs with no problems:
```
for s1 in combinations(@['A','B','C','D'], 2):
  echo s1
```
printing
```
@[A, B]
@[A, C]
@[A, D]
@[B, C]
@[B, D]
@[C, D]
```

### LINKED ISSUE #14165: Invalid C code for inline iterator with inner proc
While trying out the different nim solutions to the Hamming numbers task from RosettaCode I found that one variant did not compile. Reducing it to the max gave me this:

### 
```nim
iterator log_nodups_hamming(): int {.inline.} =
  let lb3 = 1
  proc mul3(): int = lb3
  yield mul3()

for h in log_nodups_hamming():
  break
for h in log_nodups_hamming():
  break
```

### Output
```
/mnt/space/data/nim/.cache/hn_5_test/d/@mhn_5_test.nim.c: In function ‘NimMainModule’:
/mnt/space/data/nim/.cache/hn_5_test/d/@mhn_5_test.nim.c:214:5: error: ‘colonenv_’ undeclared (first use in this function); did you mean ‘colonenv__2’?
  214 |   (*colonenv_).lb31 = ((NI) 1);
      |     ^~~~~~~~~
      |     colonenv__2
/mnt/space/data/nim/.cache/hn_5_test/d/@mhn_5_test.nim.c:214:5: note: each undeclared identifier is reported only once for each function it appears in
```
This is the expansion of the second loop.

### Additional Information

* Original source: https://rosettacode.org/wiki/Hamming_numbers#Imperative_iterator_implementation_of_the_above_functional_version
* Using `--gc:arc` the compilation succeeds.

#### Used versions
```
$ nim -v
Nim Compiler Version 1.2.0 [Linux: amd64]
Compiled at 2020-04-03
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```

* The 1.2.1 nightly from 28.4. and the 1.3.1 nightly from 29.4. show the same problem.

### LINKED ISSUE #18739: C codegen error (error: ‘colonenv_’ undeclared) when using same iterator twice
C codegen error (error: ‘colonenv_’ undeclared) when using same iterator twice

happens both in stable & devel

### Example
comment last line to compile successfully

https://play.nim-lang.org/#ix=3wLt
```nim
import std/[
  sequtils
]

type
  VisitOrder = enum
    voLeaf,
    voRoot
  MyNode = object
    name: string
    children: seq[MyNode]
    visited: bool


iterator walk(node: MyNode, visitOrder = voLeaf): MyNode =
  ## Walk tree, level order, direction by visitOrder parameter

  var resultSeq = newSeq[MyNode]()

  func callback(node: MyNode) =
    resultSeq.add node

  func visit(node: var MyNode) =
    callback(node)
    node.visited = true

  func resetVisited(node: var MyNode) =
    node.visited = false
    for c in node.children.mitems:
      c.resetVisited

  func allVisited(node: MyNode): bool =
    result = node.visited
    for c in node.children:
      result = result and c.allVisited

  func leafOrder(node: var MyNode) =
    if not node.visited and node.children.allIt(it.visited): # is leaf
      node.visit
    else:
      for c in node.children.mitems:
        c.leafOrder

  func rootOrder(node: var MyNode) =
    func aux(node: var MyNode) =
      for c in node.children.mitems:
        c.visit
      for c in node.children.mitems:
        c.aux
    node.visit
    node.aux

  var mutableMyNode = node
  mutableMyNode.resetVisited
  while not mutableMyNode.allVisited:
    case visitOrder
    of voLeaf: mutableMyNode.leafOrder
    of voRoot: mutableMyNode.rootOrder
  
  for n in resultSeq:
    yield n


if isMainModule:
  let root = MyNode(
    name: "A1",
    children: @[
      MyNode(
        name: "B1",
        children: @[
          MyNode(name: "C1")
        ]
      ),
      MyNode(
        name: "B2",
        children: @[
          MyNode(name: "C2"),
          MyNode(
            name: "C3",
            children: @[
              MyNode(name: "D1")
            ]
          )
        ]
      )
    ]
  )

  for n in root.walk(voLeaf): echo n
  for n in root.walk(voLeaf): echo n # commenting this line avoids C codegen error
```

### Current Output
```
CC: experiments.nim
/home/arkanoid/snap/codium/common/.cache/nim/experiments_d/@mexperiments.nim.c: In function ‘NimMainModule’:
/home/arkanoid/snap/codium/common/.cache/nim/experiments_d/@mexperiments.nim.c:1417:25: error: ‘colonenv_’ undeclared (first use in this function); did you mean ‘colonenv__2’?
 1417 |    asgnRef((void**) (&(*colonenv_).resultSeq1), newSeq_experiments_19(((NI) 0)));
      |                         ^~~~~~~~~
      |                         colonenv__2
/home/arkanoid/snap/codium/common/.cache/nim/experiments_d/@mexperiments.nim.c:1417:25: note: each undeclared identifier is reported only once for each function it appears in
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I'/home/arkanoid/.choosenim/toolchains/nim-#devel/lib' -I/home/arkanoid/nim/src -o /home/arkanoid/snap/codium/common/.cache/nim/experiments_d/@mexperiments.nim.c.o /home/arkanoid/snap/codium/common/.cache/nim/experiments_d/@mexperiments.nim.c' failed with exit code: 1
```

### Expected Output
```
compilation successful
```

```
nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-08-23
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 2d84f6e7c836bf18c0ecbe07fb413166861196ac
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ringabout**
fix the second example of #6269 
fixes #14165 
fixes #18739

### CODE REVIEW

File: compiler/transf.nim [hunk 1]
```diff
@@ -734,7 +734,7 @@ proc transformFor(c: PTransf, n: PNode): PNode =
       stmtList.add(newAsgnStmt(c, nkFastAsgn, temp, arg))
       idNodeTablePut(newC.mapping, formal, temp)
 
-  let body = transformBody(c.graph, c.idgen, iter, true)
+  let body = transformBody(c.graph, c.idgen, iter, true).copyTree
```
Comment (Araq) [inline compiler/transf.nim:737]: Seem expensive and worse, weird. How come for example that not a fresh set of local variables is required?

Comment (ringabout) [reply compiler/transf.nim:737]: It is a regression: https://github.com/nim-lang/Nim/commit/b1ff37c08f80b440e5ca0e4281a8967396efe7a5

Ref https://github.com/nim-lang/Nim/issues/14165#issuecomment-621641793

Comment (ringabout) [reply compiler/transf.nim:737]: Before 0.20.0, copyTree is used.Then the optimization may remove it wrongly. 
copyTree is introduced in here: https://github.com/nim-lang/Nim/commit/5432b2b83aebb139324cf44d1ebd4a88a7c85e39#diff-c7b80f51fb685eb22c5b56ee2f320d6c708706f3ae7293478ecd104a2b5b8096

**Comment by Araq** [comment]
Alternative is here https://github.com/nim-lang/Nim/pull/18812

**Pull request closed without merging**